### ISSUE TITLE
`readability-implicit-bool-conversion` false positive when `!!` is used

### ISSUE THREAD

**Issue by firewave** [llvm/llvm-project#54924]
```cpp
int cb();

bool f()
{
	return !!cb();
}
```

```
<source>:5:11: warning: implicit conversion 'int' -> bool [readability-implicit-bool-conversion]
        return !!cb();
                ~^
                (     == 0)
1 warning generated.
```

https://godbolt.org/z/zrddreEeY

**Comment by creator1creeper1** [comment]
I'm not a `clang-tidy` contributor, but this looks like a correct positive to me. `!cb()` is clearly an implicit conversion from int to bool.

**Comment by AaronBallman** [comment]
This looks like a defensible true positive but it also is a sufficiently common coding pattern for converting to bool without using a cast that I could see an argument for allowing it.

https://codesearch.isocpp.org/cgi-bin/cgi_ppsearch?q=%21%21&search=Search

**Comment by creator1creeper1** [comment]
IMO rewriting this as `cb() != 0` is perfectly reasonable.

**Comment by AaronBallman** [comment]
Personally, I think `!!` is a clever trivia question answer and rewriting to `!= 0` is a more clear approach. That said, I wouldn't be opposed if someone made an option to support `!!` given that it does have some prevalence. (I'd recommend it should be off by default though, as this check is about code readability.)

**Comment by firewave** [comment]
MSVC has a warning about this - see https://docs.microsoft.com/en-us/cpp/error-messages/compiler-warnings/compiler-warning-level-3-c4800. 
```
<source>(5): warning C4800: Implicit conversion from 'int' to bool. Possible information loss
<source>(5): note: consider using explicit cast or comparison to 0 to avoid this warning
```
https://godbolt.org/z/vrW6Yj5dj

Although it does suggest the `!= 0` change (it did not do that in the past) using `!!` was a more elegant solution which also gets rid of the warning.

**Comment by carlosgalvezp** [comment]
IMHO the `!!` pattern hinders readability, and as such it would be strange to add support for it in a `readability` check. The solution is trivial and makes the code more readable.

**Comment by SimplyDanny** [comment]
This thread is in an unclear state now. There are two contradictory suggestions:

1. Add an option to accept `!!`.
2. Do nothing.

@AaronBallman and @carlosgalvezp: You brought up these solutions. Can you live with either of them? I tend to do nothing (2), but I'm also open to implement the option.

**Comment by PiotrZSL** [comment]
I agree with Carlos. I do not think that changing a check to add option to "support" such notation is worth a time.

**Comment by AaronBallman** [comment]
I think that's reasonable; I'm going to close the issue as Not Planned, but feel free to reopen in the future if there's new information we should be aware of.
